src/web/lib/braze/LocalMessageCache.ts

```diff
@@ -0,0 +1,76 @@
+import type appboy from '@braze/web-sdk-core';
+import { SlotName } from './BrazeMessages';
+
+const localStorageKeyBase = 'gu.brazeMessageCache';
+
+type Message = appboy.InAppMessage;
+
+interface Store {
+	getItem: (key: string) => string | null;
+	setItem: (key: string, value: string) => void;
+	removeItem: (key: string) => void;
+}
+
+const keyFromSlotName = (slotName: SlotName): string =>
+	`${localStorageKeyBase}.${slotName}`;
+
+class LocalMessageCache {
+	store: Store;
+
+	constructor(store: Store) {
+		this.store = store;
+	}
+
+	shift(slotName: SlotName): Message | undefined {
+		const key = keyFromSlotName(slotName);
+		const queue = this.readQueue(key);
+		const topItem = queue.shift();
+
+		if (topItem) {
+			this.store.setItem(key, JSON.stringify(queue));
+			return topItem;
+		}
+	}
+
+	peek(slotName: SlotName): Message | undefined {
+		const key = keyFromSlotName(slotName);
+		const queue = this.readQueue(key);
+		const topItem = queue.shift();
+
+		if (topItem) {
+			return topItem;
+		}
+	}
+
+	push(slotName: SlotName, message: Message) {
+		const key = keyFromSlotName(slotName);
+		const queue = this.readQueue(key);
+
+		queue.push(message);
+		this.store.setItem(key, JSON.stringify(queue));
+	}
+
+	clear() {
+		const allSlots: SlotName[] = ['Banner', 'EndOfArticle'];
```

Comment:
Does this line have exhaustivity checking from the type-system? As it stands, I think it would be easy for us to add a new `SlotName` in the file above, but forget to include it here. This would mean we're failing to clear Braze message data for one of the queues. This might be a reason to favour a single queue, as mentioned in a previous comment.

Reply:
No, but you're right that it should. I'm pretty sure this is possibly with enum types. I'll investigate.

Reply:
I've made this change, using a TypeScript enum. There are a couple of annoying TS things to get it working so it's not as clean as I'd like, but it gets the job done.